### PR TITLE
Transaction: add toLegacyBuffer

### PR DESCRIPTION
**Pull request by afk11**
Adds `Transaction.getLegacyBuffer`

Related issue: https://github.com/bitcoinjs/bitcoinjs-lib/issues/688

### CODE REVIEW

**Comment by dcousens** [comment]
> getWitnessId returns a Buffer - it could have been hex, but since a witness-hash isn't a thing..

That is too confusing IMHO,  its almost not worth having it if it is that confusing.
Separate PR?

**Comment by afk11** [comment]
Yeah getId kind of sets the convention for returning hex. I'll axe getWitnessId until it's discussed further, though the wtxid is used in the witness coinbase commitment.

File: test/transaction.js [hunk 1]
```diff
@@ -56,13 +56,29 @@ describe('Transaction', function () {
         var actual = Transaction.fromHex(txHex)
 
         assert.strictEqual(actual.toHex(), txHex)
+        if (!f.whex) {
+          // fromHex tolerates either format, so only test this if witness is not set
+          assert.strictEqual(actual.toLegacyBuffer().toString('hex'), txHex)
+        }
+        if (f.id) {
+          // id is always whatever's in the file
```
Comment (dcousens) [inline test/transaction.js]: what do you mean in the file?

**Comment by dcousens** [comment]
LGTM

File: test/transaction.js (hunk 1 above)
Comment (dcousens) [inline test/transaction.js]: so why do `if (f.id)` at all?

Comment (afk11) [inline test/transaction.js]: Actually, removing that one since the getId test will test the witness if available.

**Comment by dcousens** [comment]
ACK, will merge on travis pass

**Comment by dcousens** [comment]
rebased, will merge

**Comment by dcousens** [comment]
@afk11 for documentation,  could you list the reasons why we need this?

**Comment by dcousens** [comment]
So,  to list the reason from: https://github.com/bitcoin/bitcoin/pull/9194

* Downstream users (not users of this library) may not support SegWit yet,  so this is to allow clients to pass them `txHex` that may be defaulting to the newer format.

I'm inclined to think this change is nice to have, but really should be discouraged by putting the burden on the downstream users to upgrade instead.

Perhaps we'll keep the PR open for as long as until some complains/asks about it?

**Comment by dcousens** [comment]
> Maybe we should work out better names for toBuffer / toLegacyBuffer, since it's only possible to have a witness serialization if one of the outputs called for it?

You mean,  if one of the inputs (previous outputs) required it?
`toBuffer` is meant to represent the default serialization,  on the blockchain.

If this libraries assumes segregated witness is in use (it will in `3.0.0`),  then,  that is representative.

The only reason to use `toLegacyBuffer`,  is supporting non-upgraded clients... AFAIK.
I think my point in https://github.com/bitcoinjs/bitcoinjs-lib/pull/715#issuecomment-267202804 holds best,  we don't merge this unless someone asks for it?

**Comment by afk11** [comment]
Agree, we can hold this one off until someone calls for it

**Comment by dcousens** [comment]
Closing (but keeping (not deleting) the branch)

**Pull request closed without merging**